File: golang/linkedlist/reverse-linked-list/reverse_list.go
package reverse_linked_list

//  反转列表

// ListNode  Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

// ReverseList 1. 双指针解法
func ReverseList(head *ListNode) *ListNode {
	var pre *ListNode
	cur := head
	for cur != nil {
		temp := cur.Next
		cur.Next = pre
		pre = cur
		cur = temp
	}
	return pre
}

// ReverseList 2. 递归解法
//func ReverseList(head *ListNode) *ListNode {
//	return help(nil, head)
//}
//
//func help(pre, head *ListNode) *ListNode {
//	if head == nil {
//		return pre
//	}
//	temp := head.Next
//	head.Next = pre
//	return help(head, temp)
//}

File: typescript/src/linkedlist/reverse-linked-list/reverseList.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

export function reverseList(head: ListNode | null): ListNode | null {
    let pre: ListNode | null = null
    let cur: ListNode | null = head
    while (cur) {
        let temp = cur.next
        cur.next = pre
        pre = cur
        cur = temp
    }
    return  pre
};